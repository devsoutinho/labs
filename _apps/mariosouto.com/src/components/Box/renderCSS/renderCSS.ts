import { StyleSheet } from '@src/theme/StyleSheet';
import themeBase, { Theme } from '@src/theme/theme';

function parseStyleSheetRule(property, value) {
  if(property.includes('Vertical')) {
    const propertyBase = property.replace('Vertical', '');
    return {
      [`${propertyBase}-top`]: value,
      [`${propertyBase}-bottom`]: value,
    }    
  }
  if(property.includes('Horizontal')) {
    const propertyBase = property.replace('Horizontal', '');
    return {
      [`${propertyBase}-left`]: value,
      [`${propertyBase}-right`]: value,
    }    
  }

  return {
    [property]: value,
  }
}

function parseStyleSheetValue({styleSheetKey, mediaResultBy}) {
  const styleSheetProperty = styleSheetKey;
  return (acc, [mediaBase, valueBase]) => {

    console.log(styleSheetKey);
  
    const media = mediaResultBy[mediaBase];
    const value = valueBase;
    return {
      ...acc,
      [media]: parseStyleSheetRule(styleSheetProperty, value),
    }
  }
}

function parseStyleSheet({ mediaResultBy, stateResultBy }) {
  return (acc, [styleSheetKey, styleSheetValue]) => {
    if(stateResultBy[styleSheetKey]) {
      const state = stateResultBy[styleSheetKey];
      const styleSheet = styleSheetValue;
      const output = Object.entries(styleSheet).reduce(parseStyleSheet({ mediaResultBy, stateResultBy }), {});
      return {
        ...acc,
        [state]: output
      }
    }

    if(typeof styleSheetValue === 'string' || typeof styleSheetValue === 'number') {
      const key = styleSheetKey;
      const value = styleSheetValue
      return { ...acc, ...parseStyleSheetRule(key, value) };
    }
  
    if(typeof styleSheetValue === 'object') {
      return {
        ...acc,
        ...Object.entries(styleSheetValue).reduce(parseStyleSheetValue({styleSheetKey, mediaResultBy}), {}),
      }
    }
    
  
    return { ...acc }
  }
}

export function renderCSS(styleSheet: StyleSheet = {}, theme: Theme = themeBase, uniqueId: string = 'demo-id') {
  const breakpoints = theme.breakpoints;
  const mediaResultBy = {
    xs: `@media (min-width: ${breakpoints.xs}px)`,
    sm: `@media (min-width: ${breakpoints.sm}px)`,
    md: `@media (min-width: ${breakpoints.md}px)`,
    lg: `@media (min-width: ${breakpoints.lg}px)`,
    xl: `@media (min-width: ${breakpoints.xl}px)`,
  };

  const stateResultBy = {
    hover: `&:hover, body.dark &.${uniqueId}:hover`,
    focus: `&:focus, body.dark &.${uniqueId}:focus`,
    disabled: `&:disabled, body.dark &.${uniqueId}:disabled`,
    modeDark: `body.dark &.${uniqueId}`,
  };

  const output = Object.entries(styleSheet).reduce(parseStyleSheet({ mediaResultBy, stateResultBy }), {});
  return output;
}
