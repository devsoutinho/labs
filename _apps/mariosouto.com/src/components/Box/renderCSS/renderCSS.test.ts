import { renderCSS } from './renderCSS';

describe('renderCSS()', () => {
  describe('when receives a stylesheet with basic styles', () => {
    it('returns it as expected', () => {
      expect(renderCSS({ color: 'red' })).toEqual({ color: 'red' });
    });
    it('if they are responsive, return as expected', () => {
      expect(renderCSS({ color: { xs: 'red', md: 'blue' } })).toMatchSnapshot();
    })
  });
  describe('when receives a stylesheet with state styles', () => {
    it('returns it as expected', () => {
      expect(renderCSS({ color: 'red', hover: { color: 'blue' } })).toMatchSnapshot();
    });
    it('if they are responsive, return as expected', () => {
      expect(renderCSS({ color: 'red', hover: { color: { xs: 'blue' } } })).toMatchSnapshot();
    })
  });
});
