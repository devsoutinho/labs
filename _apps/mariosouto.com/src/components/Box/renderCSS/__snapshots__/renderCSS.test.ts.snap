// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderCSS() when receives a stylesheet with basic styles if they are responsive, return as expected 1`] = `
Object {
  "@media (min-width: 0px)": Object {
    "color": "red",
  },
  "@media (min-width: 768px)": Object {
    "color": "blue",
  },
}
`;

exports[`renderCSS() when receives a stylesheet with hover styles if they are responsive, return as expected 1`] = `
Object {
  "body &:hover": Object {
    "@media (min-width: 0px)": Object {
      "color": "blue",
    },
  },
  "color": "red",
}
`;

exports[`renderCSS() when receives a stylesheet with hover styles returns it as expected 1`] = `
Object {
  "body &:hover": Object {
    "color": "blue",
  },
  "color": "red",
}
`;
